docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,279 @@
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type sbom/example localhost:5000/sample/images:foobar
+Error: neither file nor annotation provided in the command
+Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<type>] [...]
+To attach to an existing artifact, please provide files via argument or annotations via flag "--annotation". Run "oras attach -h" for more options and examples
+```
+
+#### Example 7: When pushing files, if the annotation file doesn't match the required format
+
+Current behavior and output:
+
+```
+$ oras push --annotation-file sbom.json ghcr.io/library/alpine:3.9
+Error: failed to load annotations from sbom.json: json: cannot unmarshal string into Go value of type map[string]map[string]string. Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation-file annotation.json ghcr.io/library/alpine:3.9
+Error: invalid annotation json file: failed to load annotations from annotation.json.
+Annotation file doesn't match the required format. Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations
+```
+
+#### Example 8: When pushing files, if the annotation value doesn't match the required syntax
+
+Current behavior and output:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: missing key in `--annotation` flag: key:value
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: annotation value doesn't match the required format.
+Try `oras push --annotation "key=value" ghcr.io/library/alpine:3.9`  
+```
+
+#### Example 9: When failed to pull files from a public registry
+
+```
+$ oras pull docker.io/nginx:latest
+Error: failed to resolve latest: GET "https://registry-1.docker.io/v2/nginx/manifests/latest": response status code 401: unauthorized: authentication required: [map[Action:pull Class: Name:nginx Type:repository]]
+```
+
+Suggested error message:
+
+```
+$ oras pull docker.io/nginx:latest
+Error response from registry: pull access denied for docker.io/nginx:latest : response status code 401: unauthorized: requested access to the resource is denied
+Namespace is missing, do you mean `oras pull docker.io/library/nginx:latest`? 
+```
+
+#### Example 10: Neither registry nor OCI image layout provided when pushing a folder 
+
+Current behavior and output:
+
+```
+$ oras push /oras --format json
+Error: Head "https:///v2/oras/manifests/sha256:ffa50b27cd0096150c0338779c5090db41ba50d01179d851d68afa50b393c3a3": http: no Host in request URL
+```
+
+Suggested error message:
+
+```
+$ oras push /oras --format json
+Error: neither registry nor OCI image layout provided in the command 
+Usage: oras push [flags] <name>[:<tag>[,<tag>][...]] <file>[:<type>] [...]
+please specify a registry or an OCI image layout folder to push. Run "oras push -h" for more options and examples 
```

Comment:
Since user haven't use `--oci-layout` here, the provided reference `/oras` is expected to be an artifact in a remote registry. To avoid confusion, suggest not to mention OCI layout.
```suggestion
Error: "/oras" is an invalid reference
Usage: oras push [flags] <name>[:<tag>[,<tag>][...]] <file>[:<type>] [...]
Please specify a valid reference in the form of <registry>/<repo>[:tag|@digest]
```